### PR TITLE
Add Path chooser. Add DiskMonitort multiple disks support

### PR DESCRIPTION
**Pull request by ObjatieGroba**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature
- Bug Fix
- Improvement
- Performance Improvement
- Backward Incompatible Change
- Build/Testing/Packaging Improvement
- Other

Short description (up to few sentences):

...

Detailed description (optional):

...

### CODE REVIEW

File: dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 1]
```diff
@@ -513,7 +522,10 @@ MergeTreeData::MutableDataPartPtr MergeTreeDataMergerMutator::mergePartsToTempor
               << parts.front()->name << " to " << parts.back()->name
               << " into " << TMP_PREFIX + future_part.name);
 
-    String new_part_tmp_path = data.getFullPath() + TMP_PREFIX + future_part.name + "/";
+    size_t expected_size = future_part.expectedSize();
+    String part_path = data.getFullPathForPart(expected_size); ///@TODO_IGR ASK EXPECTED SIZE
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp]: need constant

File: dbms/src/Interpreters/Context.cpp [hunk 2]
```diff
@@ -1621,6 +1623,23 @@ CompressionCodecPtr Context::chooseCompressionCodec(size_t part_size, double par
 }
 
 
+///@TODO_IGR ASK maybe pointer to Schema?
+Schema Context::chooseSchema(const String & name) const
```
Comment (alesapin) [inline dbms/src/Interpreters/Context.cpp]: Why not const reference?

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h [hunk 3]
```diff
@@ -118,32 +149,125 @@ class DiskSpaceMonitor
...
+    ///@TODO_IGR REMOVE it
+    size_t size() const {
+        return disks.size();
+    }
+
+private:
+    std::map<String, Disk> disks;
+};
+
+class Schema
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Too common name. MergeTreeStorageSchema or DiskSchema?

File: dbms/src/Interpreters/Context.cpp (hunk 2 above)
Comment (alesapin) [reply dbms/src/Interpreters/Context.cpp]: Why `choose`? We don't choose anything here. Just `get` element by key.

File: dbms/src/Storages/MergeTree/ActiveDataPartSet.h [hunk 4]
```diff
@@ -15,8 +15,16 @@ namespace DB
 class ActiveDataPartSet
 {
 public:
+    struct PartPathName {
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/ActiveDataPartSet.h]: `PartPathWithName` or something like this. Also not obvious why we need separate class and cannot get name with `basename` and path with `dirname`.

File: dbms/src/Storages/MergeTree/DataPartsExchange.cpp [hunk 5]
```diff
@@ -200,7 +200,9 @@ MergeTreeData::MutableDataPartPtr Fetcher::fetchPart(
     String tmp_prefix = tmp_prefix_.empty() ? TMP_PREFIX : tmp_prefix_;
 
     String relative_part_path = String(to_detached ? "detached/" : "") + tmp_prefix + part_name;
-    String absolute_part_path = data.getFullPath() + relative_part_path + "/";
+    auto reservation = data.reserveSpaceForPart(0); ///@TODO_IGR ASK What size should be there?
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DataPartsExchange.cpp]: I think we have to add new http handler (or params to existing handler) which will tell about part size.

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h [hunk 6]
```diff
@@ -118,32 +149,125 @@ class DiskSpaceMonitor
...
+class MergeTreeStorageConfiguration {
+public:
+    MergeTreeStorageConfiguration(const Poco::Util::AbstractConfiguration & config, const std::string & config_prefix);
+
+    const Schema & operator[](const String & name) const {
+        return schema_selector[name];
+    }
+
+private:
+    DisksSelector disks;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Why we need it?

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h [hunk 7]
```diff
@@ -118,32 +149,125 @@ class DiskSpaceMonitor
...
+        return disks.size();
+    }
+
+private:
+    std::map<String, Disk> disks;
+};
+
+class Schema
+{
+    class Volume {
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Unclear name.

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.cpp [hunk 8]
```diff
@@ -3,8 +3,184 @@
...
+        disks.push_back(disk_selector[disk_name]);
+    }
+}
+
+bool Schema::Volume::setDefaultPath(const String & path) {
+    bool set = false;
+    for (auto & disk : disks) {
+        if (disk.path == "default") {
+            if (set) {
+                throw Exception("It is not possible to have two default disks", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG); ///@TODO_IGR ASK ErrorCode
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.cpp]: Validate on config parsing stage.

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h [hunk 9]
```diff
@@ -118,32 +149,125 @@ class DiskSpaceMonitor
         return res;
     }
 
-    static UInt64 getReservedSpace()
+    static UInt64 getAllReservedSpace()
     {
         std::lock_guard lock(mutex);
-        return reserved_bytes;
+        UInt64 res;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Undefined behaviour

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h [hunk 10]
```diff
@@ -118,32 +149,125 @@ class DiskSpaceMonitor
...
+        }
+        return res;
     }
 
-    static UInt64 getReservationCount()
+    static UInt64 getAllReservationCount()
     {
         std::lock_guard lock(mutex);
-        return reservation_count;
+        UInt64 res;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Undefined behaviour

File: dbms/src/Storages/MergeTree/MergeTreeData.cpp [hunk 11]
```diff
@@ -693,10 +708,12 @@ void MergeTreeData::loadDataParts(bool skip_sanity_checks)
...
                 size_t contained_parts = 0;
 
-                LOG_ERROR(log, "Part " << full_path + file_name << " is broken. Looking for parts to replace it.");
+                LOG_ERROR(log, "Part " << full_paths[path_index] + file_name << " is broken. Looking for parts to replace it.");
 
-                for (const String & contained_name : part_file_names)
+                for (auto part_file_name : part_file_names)
                 {
+                    const String & contained_name = part_file_name.first;
+                    const size_t contained_path_index = part_file_name.second;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeData.cpp]: It's not part index, it is index in full paths array.

File: dbms/src/Storages/MergeTree/MergeTreeDataPart.h [hunk 12]
```diff
@@ -86,6 +86,7 @@ struct MergeTreeDataPart
 
     const MergeTreeData & storage;
 
+    String path;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeDataPart.h]: Bad name. We have bunch of different paths.

File: dbms/src/Storages/MergeTree/MergeTreeData.h [hunk 13]
```diff
@@ -569,6 +569,17 @@ class MergeTreeData : public ITableDeclaration
     MergeTreeData::MutableDataPartPtr cloneAndLoadDataPart(const MergeTreeData::DataPartPtr & src_part, const String & tmp_part_prefix,
                                                            const MergeTreePartInfo & dst_part_info);
 
+    DiskSpaceMonitor::ReservationPtr reserveSpaceAtDisk(UInt64 expected_size) const; ///@TODO_IGR ASK Maybe set this method as private?
+
+    Strings getFullPaths() const {
+        auto paths = schema.getFullPaths();
+        for (auto && path : paths) {
+            path += table_name + '/'; ///@TODO_IGR ASK It is too slow(
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeData.h]: Why do we need it?

File: dbms/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp [hunk 14]
```diff
@@ -233,7 +233,7 @@ void ReplicatedMergeTreePartCheckThread::checkPart(const String & part_name)
                 zk_part_header.getChecksums().checkEqual(local_part_header.getChecksums(), true);
 
                 checkDataPart(
-                    storage.data.getFullPath() + part_name,
+                    storage.data.getFullPaths()[0] + part_name, ///@TODO_IGR ASK what should we do there? Should we check all paths in checkDataPart?
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp]: No, we should check part `part_path`.

File: dbms/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp [hunk 15]
```diff
@@ -32,8 +32,8 @@ void ReplicatedMergeTreeQueue::addVirtualParts(const MergeTreeData::DataParts &
 
     for (const auto & part : parts)
     {
-        current_parts.add(part->name);
-        virtual_parts.add(part->name);
+        current_parts.add("/", part->name);
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp]: What purpose of this "/"? Seems like wrong usage.

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 16]
```diff
@@ -333,7 +339,8 @@ struct CurrentlyMergingPartsTagger
 
 void StorageMergeTree::mutate(const MutationCommands & commands, const Context &)
 {
-    MergeTreeMutationEntry entry(commands, full_path, data.insert_increment.get());
+    const auto full_paths = data.getFullPaths(); ///@TODO_IGR ASK What expected size of mutated part? what size should we reserve?
+    MergeTreeMutationEntry entry(commands, full_paths[0], data.insert_increment.get()); ///@TODO_IGR ASK PATH TO ENTRY
```
Comment (alesapin) [inline dbms/src/Storages/StorageMergeTree.cpp]: As you decide

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 17]
```diff
@@ -227,6 +226,9 @@ StorageReplicatedMergeTree::StorageReplicatedMergeTree(
     if (path_.empty())
         throw Exception("ReplicatedMergeTree storages require data path", ErrorCodes::INCORRECT_FILE_NAME);
 
+    ///@TODO_IGR ASK Set inside MergeTreeData?
+    data.schema.setDefaultPath(path_);
```
Comment (alesapin) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: Yes, better in constructor.

File: dbms/src/Storages/System/StorageSystemTables.cpp [hunk 18]
```diff
@@ -195,8 +195,11 @@ class TablesBlockInputStream : public IBlockInputStream
                 if (columns_mask[src_index++])
                     res_columns[res_index++]->insert(0u);  // is_temporary
 
-                if (columns_mask[src_index++])
-                    res_columns[res_index++]->insert(tables_it->table()->getDataPath());
+                if (columns_mask[src_index++]) {
+                    for (const String & path : tables_it->table()->getDataPaths() ) {
+                        res_columns[res_index++]->insert(path);  ///@TODO_IGR ASK Is it fine?
```
Comment (alesapin) [inline dbms/src/Storages/System/StorageSystemTables.cpp]: If it's full path, than it's ok.

**Review by alesapin — COMMENTED** [review]
Need tests and example of usage in documentation.

File: dbms/src/Storages/MergeTree/DiskSpaceMonitor.h (hunk 3 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/DiskSpaceMonitor.h]: Shema and Volume would be renamed.
Nobody knows fine names for this classes

File: dbms/src/Storages/MergeTree/ActiveDataPartSet.h (hunk 4 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/ActiveDataPartSet.h]: Class ActiveDataPartSet? Or struct PartPathWithNames?

File: dbms/src/Storages/MergeTree/MergeTreeDataPart.h (hunk 12 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/MergeTreeDataPart.h]: Hm, 
path + relative_path + name = full_path
What name should we use for path?
Can you suggest?

File: dbms/src/Storages/MergeTree/MergeTreeData.h (hunk 13 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/MergeTreeData.h]: To iterate over all paths to
1) Remove all perts
2) Load all parts from disk

File: dbms/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp (hunk 14 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp]: How can we get part_path by part_name?
Or we shold chg queu by containig path in queue

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 16 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/StorageMergeTree.cpp]: What about expected size for such entry?

File: dbms/src/Storages/MergeTree/MergeTreeDataPart.h (hunk 12 above)
Comment (ObjatieGroba) [reply dbms/src/Storages/MergeTree/MergeTreeDataPart.h]: data_path + relative_path + name.
Mark it as private
add functions:
getFullPath() -> data_path + relative_path + name
getName() -> name
setRelativePath()

**Pull request merged**